src/server/builder.ts

```diff
@@ -0,0 +1,417 @@
+                this.setFileInfo(fileName, new BuilderFileInfo(scriptInfo, this.project));
+            }
+            return this.getFileInfo(fileName);
+        }
+
+        getFilesAffectedBy(fileName: string): string[] {
+            const info = this.getOrCreateFileInfo(fileName);
+            let result: string[];
+            if (info.checkIfShapeSignatureChanged()) {
+                const options = this.project.getCompilerOptions();
+                if (options && (options.out || options.outFile)) {
+                    result = [fileName];
+                }
+                else {
+                    result = this.project.getFileNamesWithoutDefaultLib();
+                }
+            }
+            else {
+                result = [fileName];
+            }
+            return result;
+        }
+    }
+
+    class ModuleBuilderFileInfo extends BuilderFileInfo {
+        references: ModuleBuilderFileInfo[] = [];
+        referencedBy: ModuleBuilderFileInfo[] = [];
+        scriptVersionForReferences: string;
+
+        private finalized = false;
+
+        getReferencedByFileNames() {
+            return map(this.referencedBy, info => info.scriptInfo.fileName);
+        }
+
+        static compareFileInfos(lf: ModuleBuilderFileInfo, rf: ModuleBuilderFileInfo): number {
+            const l = lf.scriptInfo.fileName;
+            const r = rf.scriptInfo.fileName;
+            return (l < r ? -1 : (l > r ? 1 : 0));
+        };
+
+        static addToReferenceList(array: ModuleBuilderFileInfo[], fileInfo: ModuleBuilderFileInfo, afterFinalized: boolean) {
+            if (!afterFinalized) {
+                array.push(fileInfo);
+            }
+            else {
+                const insertIndex = binarySearch(array, fileInfo, ModuleBuilderFileInfo.compareFileInfos);
+                if (insertIndex < 0) {
+                    array.splice(~insertIndex, 0, fileInfo);
+                }
+            }
+        }
+
+        static removeFromReferenceList(array: ModuleBuilderFileInfo[], fileInfo: ModuleBuilderFileInfo, afterFinalized?: boolean) {
+            if (!array) {
+                return;
+            }
+
+            if (!afterFinalized) {
+                // Before finalized the reference lists are not sorted
+                for (let i = 0; i < array.length; i++) {
+                    if (ModuleBuilderFileInfo.compareFileInfos(array[i], fileInfo) === 0) {
+                        array.splice(i, 1);
+                        break;
+                    }
+                }
+            }
+            else {
+                const removeIndex = binarySearch(array, fileInfo, ModuleBuilderFileInfo.compareFileInfos);
+                if (removeIndex >= 0) {
+                    array.splice(removeIndex, 1);
+                }
+            }
+        }
+
+        addReferencedBy(fileInfo: ModuleBuilderFileInfo): void {
+            if (this.referencedBy.length === 0) {

```

Comment:
nit: can you move this check to `addToReferenceList` so it will do `push` if either `afterFinalized` is `true` or `array` is empty? the code will look more consistent.